ompi/attribute/attribute.c

```diff
@@ -283,104 +282,103 @@
-        } \
-        /* MPI-2 Fortran-style */ \
-        else { \
-            MPI_Aint in, out;                        \
-            MPI_Fint attr_##type##_f;                \
-            in = translate_to_aint(in_attr); \
-            attr_##type##_f = OMPI_INT_2_FINT(((ompi_##type##_t *)old_object)->attr_##type##_f); \
-            (*((keyval_obj->copy_attr_fn).attr_aint_copy_fn)) \
-                (&attr_##type##_f, \
-                 &f_key, &keyval_obj->extra_state.f_address, &in, &out, \
-                 &f_flag, &f_err); \
-            if (MPI_SUCCESS != OMPI_FINT_2_INT(f_err)) { \
-                err = OMPI_FINT_2_INT(f_err);           \
-            } else {                                    \
-                out_attr->av_value = (void *) out;      \
-                flag = OMPI_LOGICAL_2_INT(f_flag);      \
-            }                                           \
-        } \
-    } \
-    /* C style */ \
-    else { \
-        void *in, *out; \
-        in = translate_to_c(in_attr); \
-        if ((err = (*((keyval_obj->copy_attr_fn).attr_##type##_copy_fn)) \
-              ((ompi_##type##_t *)old_object, key, keyval_obj->extra_state.c_ptr, \
-               in, &out, &flag, (ompi_##type##_t *)(new_object))) == MPI_SUCCESS) { \
-            out_attr->av_value = out;                                   \
-        }                                                               \
-    } \
-    OPAL_THREAD_LOCK(&attribute_lock); \
-} while (0)
+#define COPY_ATTR_CALLBACKS(type, old_object, keyval_obj, in_attr, new_object, out_attr, err)    \
+    do {                                                                                         \
+        OPAL_THREAD_UNLOCK(&attribute_lock);                                                     \
+        if (0 != (keyval_obj->attr_flag & OMPI_KEYVAL_F77)) {                                    \
+            MPI_Fint f_key = OMPI_INT_2_FINT(key);                                               \
+            MPI_Fint f_err;                                                                      \
+            ompi_fortran_logical_t f_flag;                                                       \
+            /* MPI-1 Fortran-style */                                                            \
+            if (0 != (keyval_obj->attr_flag & OMPI_KEYVAL_F77_INT)) {                            \
+                MPI_Fint in, out;                                                                \
+                MPI_Fint attr_##type##_f;                                                        \
+                in = translate_to_fint(in_attr);                                                 \
+                attr_##type##_f = OMPI_INT_2_FINT(                                               \
+                    ((ompi_##type##_t *) old_object)->attr_##type##_f);                          \
+                (*((keyval_obj->copy_attr_fn).attr_fint_copy_fn))(&attr_##type##_f, &f_key,      \
```

Comment:
This i unbearable to read.

Reply:
Which? The before or the after? Because the after looks much cleaner than the before. Neither are great but that can't be helped with these massive macros.